### PR TITLE
KTOR-3832 Introduce file cache; Prevent blocking when receiving conte…

### PR DESCRIPTION
**Pull request by e5l**
Fix [KTOR-3832](https://youtrack.jetbrains.com/issue/KTOR-3832)

### CODE REVIEW

File: ktor-server/ktor-server-plugins/ktor-server-double-receive/jvm/src/io/ktor/server/plugins/doublereceive/FileCacheJvm.kt [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import java.io.*
+import java.nio.*
+import kotlin.coroutines.*
+
+internal actual class FileCache actual constructor(
+    private val body: ByteReadChannel,
+    bufferSize: Int,
+    context: CoroutineContext
+) : DoubleReceiveCache {
+    private val file = File.createTempFile("ktor-double-receive-cache", ".tmp")
```
Comment (InsanusMokrassar) [inline ktor-server/ktor-server-plugins/ktor-server-double-receive/jvm/src/io/ktor/server/plugins/doublereceive/FileCacheJvm.kt:19]: Add `deleteOnExit()`?

Comment (e5l) [reply ktor-server/ktor-server-plugins/ktor-server-double-receive/jvm/src/io/ktor/server/plugins/doublereceive/FileCacheJvm.kt:19]: I would say it should be in `call.afterFinish`.

File: ktor-server/ktor-server-plugins/ktor-server-double-receive/jvmAndNix/src/io/ktor/server/plugins/doublereceive/DoubleReceiveConfig.kt [hunk 2]
```diff
@@ -24,6 +24,13 @@ public class DoubleReceiveConfig {
      */
     public var cacheRawRequest: Boolean = true
 
+    /**
+     * Indicates if a file should be used to cache raw requests. Otherwise, requests will be stored in memory.
+     *
+     * If [cacheRawRequest] is `false` then this setting is ignored.
+     */
+    public var useFileForCache: Boolean = PlatformUtils.IS_JVM
```
Comment (olme04) [inline ktor-server/ktor-server-plugins/ktor-server-double-receive/jvmAndNix/src/io/ktor/server/plugins/doublereceive/DoubleReceiveConfig.kt:32]: isn't it better to make it configurable per request, and not via one boolean flag?
similar as with `excludeFromCache`?
So `File` cache will be used only for those requests, that can contain big body. If not, using file cache will decrease performance, as writing to File is almost always slower than writing to Memory

File: ktor-server/ktor-server-plugins/ktor-server-double-receive/jvm/src/io/ktor/server/plugins/doublereceive/FileCacheJvm.kt (hunk 1 above)
Comment (InsanusMokrassar) [reply ktor-server/ktor-server-plugins/ktor-server-double-receive/jvm/src/io/ktor/server/plugins/doublereceive/FileCacheJvm.kt:19]: Yes, that too

File: ktor-server/ktor-server-plugins/ktor-server-double-receive/jvmAndNix/src/io/ktor/server/plugins/doublereceive/DoubleReceiveConfig.kt (hunk 2 above)
Comment (e5l) [reply ktor-server/ktor-server-plugins/ktor-server-double-receive/jvmAndNix/src/io/ktor/server/plugins/doublereceive/DoubleReceiveConfig.kt:32]: You can install the plugin only for the specific route which is supposed to have a big body. I'm not sure how it should be possible to configure for a specific request.

Comment (olme04) [reply ktor-server/ktor-server-plugins/ktor-server-double-receive/jvmAndNix/src/io/ktor/server/plugins/doublereceive/DoubleReceiveConfig.kt:32]: i mean use the same declaration as with `excludeFromCache`, so like:
```kotlin
fun useFileForCache(block: (call: ApplicationCall, body: Any) -> Boolean)
```
and in this case we will be able to f.e. do `file` cache when `contentLength` is big, or for specific content type, and not to apply this config for multiple routes

Comment (rsinukov) [reply ktor-server/ktor-server-plugins/ktor-server-double-receive/jvmAndNix/src/io/ktor/server/plugins/doublereceive/DoubleReceiveConfig.kt:32]: I agree, it should be configurable per request. Also, default should be opposite, IMO.

File: ktor-server/ktor-server-plugins/ktor-server-double-receive/jvm/src/io/ktor/server/plugins/doublereceive/FileCacheJvm.kt (hunk 1 above)
Comment (rsinukov) [inline ktor-server/ktor-server-plugins/ktor-server-double-receive/jvm/src/io/ktor/server/plugins/doublereceive/FileCacheJvm.kt:19]: Do you use the same file for all requests?

**Review by rsinukov — CHANGES_REQUESTED** [review]

File: ktor-server/ktor-server-plugins/ktor-server-double-receive/jvm/src/io/ktor/server/plugins/doublereceive/FileCacheJvm.kt (hunk 1 above)
Comment (e5l) [reply ktor-server/ktor-server-plugins/ktor-server-double-receive/jvm/src/io/ktor/server/plugins/doublereceive/FileCacheJvm.kt:19]: Yep, `createTempFile` will append random number to the provided prefix

Comment (rsinukov) [reply ktor-server/ktor-server-plugins/ktor-server-double-receive/jvm/src/io/ktor/server/plugins/doublereceive/FileCacheJvm.kt:19]: oh, sorry, missed this part

File: ktor-server/ktor-server-plugins/ktor-server-double-receive/jvmAndNix/src/io/ktor/server/plugins/doublereceive/DoubleReceiveConfig.kt (hunk 2 above)
Comment (e5l) [reply ktor-server/ktor-server-plugins/ktor-server-double-receive/jvmAndNix/src/io/ktor/server/plugins/doublereceive/DoubleReceiveConfig.kt:32]: @olme04  
`fun useFileForCache(block: (call: ApplicationCall, body: Any) -> Boolean)` is actually a good idea. Thank you for the hint.

@rsinukov the file default seems to be not much slower, but much safer. I would keep it as a default, but we can discuss.

Comment (e5l) [reply ktor-server/ktor-server-plugins/ktor-server-double-receive/jvmAndNix/src/io/ktor/server/plugins/doublereceive/DoubleReceiveConfig.kt:32]: Yep. And I would even put default here:`fun useFileForCache(block: (call: ApplicationCall) -> Boolean = { true })`

**Review by rsinukov — APPROVED** [review]

**Pull request merged**